### ISSUE TITLE
Rayman 3: Broken effects, most noticeable on sparks and lighting effects

### ISSUE THREAD

**Issue by RavenMacDaddy** [dege-diosg/dgVoodoo2#90]
For transparency I'm first wrapping the game with [d3d8to9](https://github.com/crosire/d3d8to9/). Why? Because without first running the game in DX9 it stutters like crazy all the time.

With only d3d8to9 running with the game, everything is stutter free and free from graphical glitches, but if I then add dgVoodoo2 into the mix (with any output API selected), broken effects start to occur.

You can see it on fires, hit sparks, trails when collecting gems, even when enemies are moving in certain situations - anything like it.

Maybe other effects are broken too, but those are the most noticeable.

Before writing this issue, I tried lots of combinations to try and get the best from both worlds; stutter free with no broken effects while running DX11 or 12.

I tried changing INIs for d3d8to9, using different DX8 DLLs, DX9 DLLs, all of the ones in the MS/x86 folder of dgVoodoo2 - you name it.

The game version is the one from GOG with the Better Rayman 3 mod applied for better compatibility with modern systems.

The OS is Windows 11 - and for reference all of the issues described are present on Windows 10 as well and without any mods applied whatsoever as far as I know (just installing the GOG version and running it that's).

---

The goal is to get to use Auto HDR with the game, which is why DX10 and up is needed - something that your software makes possible, and I really appreciate it - thank you.

**Comment by dege-diosg** [comment]
I played around with the game a little bit more carefully, and I think I managed to reproduce the problem.
I found a location where it's relatively easy to produce flashing empty half-translucent rectangles:

![image](https://user-images.githubusercontent.com/71823288/137157401-84fc3077-0e22-4e3d-9403-415175b1f061.png)
That shiny 'orb' shouldn't be there, it's flashing across various frames for a while then disappears (Rayman is just standing, his or the camera position is not changing).
This capture is from the vanilla version of the game with native D3D8, without dgVoodoo (to keep things simple).
I could reproduce it with dgVoodoo D3D8 as well and managed to trace down the Draw call belonging to those unnecessary sparkling rectangles but honestly, I couldn't see anything strange in the API call log. In fact those sparkling rectangles are continuously drawn in multiple frames but the incoming vertex data can vary, ending up in it's either invisible or not.

Based on this, and the fact that the issue arises with both native and dgVoodoo D3D8, I think it must be a bug in the game code itself. Mismanaged vertex buffers or something like that, but it's not clear what it depends on, why it comes more easily with dgVoodoo than natively.

BetterRayman3 intercepts the D3D8 calls coming from the game and modifies them before forwarding them to either 'the real' D3D8 or D3D9, depending on the config in d3d8.ini. This is to provide more resolutions, widescreen support and such. And, the fps stutter is also fixed in that, I think.
